File: converted_source/used/Uparams.pas
unit Uparams;

interface

uses ucollect;

const
  { FieldType }
  kFieldUndefined = 0;
  kFieldFloat = 1;
  kFieldSmallint = 2;
  kFieldColor = 3;
  kFieldBoolean = 4;
  kFieldThreeDObject = 5;
  kFieldEnumeratedList = 6;
  kFieldHeader = 7;
  kFieldLongint = 8;
    {note: the longint type is NOT for use for parameters (in param panels or breeding).
     if you want to use a longint for a parameter, you have to
     update the breeding method and the method that creates param panels. the smallint panel will not
     work correctly with a longint as it is now (but you could subclass it if you need one).}

	{ IndexType - for arrays }               
  kIndexTypeUndefined = 0;
  kIndexTypeNone = 1;
  kIndexTypeSCurve = 3;

  { transferType - how data is transferred to and from model object }
  kTransferTypeUndefined = 0;
  kTransferTypeMFD = 1;  {mfd is short for MoveFieldData}
  kTransferTypeGetSetSCurve = 2;
  kTransferTypeObject3D = 3;

  kNotArray = -1;

type
  PdParameter = class
    public
    fieldNumber: smallint;
    fieldID: string[80];
    name: string[80];
    fieldType, indexType: smallint;
    unitSet, unitModel, unitMetric, unitEnglish: smallint;
    lowerBoundOriginal, upperBoundOriginal: single;
    defaultValueStringOriginal: string; {must be whole string to allow for tdos}
    cannotOverride, isOverridden: boolean;
    lowerBoundOverride, upperBoundOverride: single;
    defaultValueStringOverride: string;
    regrow, readOnly: boolean;
    accessString: string[80];
    transferType: smallint;
    hint: string;
    valueHasBeenReadForCurrentPlant: boolean;
    collapsed: boolean;
    originalSectionName: string[80]; // v2.1
    constructor create;
    constructor make(aFieldNumber: smallint; aFieldID: string; aName: string; aFieldType, anIndexType: smallint; aUnitSet, aUnitModel, aUnitMetric, aUnitEnglish: smallint;
      aLowerBound, anUpperBound: single; aDefaultValueString: string; aRegrow, aReadOnly: boolean; anAccessString: string; aTransferType: smallint; aHint: string);
    function lowerBound: single;
    function upperBound: single;
    function defaultValueString: string;
    function indexCount: integer;
    function getName: string;
    procedure setName(newName: string);
    function getHint: string;
    end;

PdParameterManager = class
 	public
  parameters: TListCollection;
  constructor create;
  destructor destroy; override;
  procedure makeParameters;
  procedure addParameter(newParameter: PdParameter);
  procedure addParameterForSection(sectionName: string; orthogonalSectionName: string; newParameter: PdParameter);
  procedure clearParameters;
  class function fieldTypeName(fieldType: integer): string;
  class function indexTypeName(indexType: integer): string;
  function parameterForIndex(index: longint): PdParameter;
  function parameterForFieldID(fieldID: string): PdParameter;
  function parameterForName(name: string): PdParameter;
  function parameterForFieldNumber(fieldNumber: longint): PdParameter;
  function parameterIndexForFieldNumber(fieldNumber: longint): integer;
  function parameterIndexForFieldID(fieldID: string): integer;
  procedure setAllReadFlagsToFalse;
  end;

implementation

uses SysUtils, Windows, Dialogs,
  uexcept, usupport, usection, uunits, umakepm, udomain;

{ ---------------------------------------------------------- PdParameter }
constructor PdParameter.create;
  begin
  collapsed := true;
  end;

constructor PdParameter.make(aFieldNumber: smallint; aFieldID: string; aName: string; aFieldType, anIndexType: smallint; aUnitSet, aUnitModel, aUnitMetric, aUnitEnglish: smallint;
    aLowerBound, anUpperBound: single; aDefaultValueString: string; aRegrow, aReadOnly: boolean; anAccessString: string; aTransferType: smallint; aHint: string);
  begin
  self.create;
  fieldNumber := aFieldNumber;
  fieldID := copy(aFieldID, 1, 80);
  name := copy(aName, 1, 80);
  fieldType := aFieldType;
  hint := aHint; // v2.0 headers can have hints
  if fieldType = kFieldHeader then exit;
  indexType := anIndexType;
  unitSet := aUnitSet;
  unitModel := aUnitModel;
  unitMetric := aUnitMetric;
  unitEnglish := aUnitEnglish;
  lowerBoundOriginal := aLowerBound;
  upperBoundOriginal := anUpperBound;
  defaultValueStringOriginal := aDefaultValueString;
  regrow := aRegrow;
  readOnly := aReadOnly;
  accessString := copy(anAccessString, 1, 80);
  transferType := aTransferType;
  //hint := aHint;
  end;

function PdParameter.indexCount: integer;
  begin
  result := 1;
  case indexType of
    kIndexTypeUndefined: result := 1;
    kIndexTypeNone: result := 1;
    kIndexTypeSCurve: result := 4;
  else
    raise Exception.create('Problem: Unexpected index type in method PdParameter.indexCount.');
  end;
  end;

function PdParameter.getName: string;
  begin
  result := self.name;
  end;

procedure PdParameter.setName(newName: string);
  begin
  name := copy(newName, 1, 80);
  end;

function PdParameter.lowerBound: single;
  begin
  if (not cannotOverride) and (isOverridden) then
    result := lowerBoundOverride
  else
    result := lowerBoundOriginal;
  end;

function PdParameter.upperBound: single;
  begin
  if (not cannotOverride) and (isOverridden) then
    result := upperBoundOverride
  else
    result := upperBoundOriginal;
  end;

function PdParameter.defaultValueString: string;
  begin
  if (not cannotOverride) and (isOverridden) then
    result := defaultValueStringOverride
  else
    result := defaultValueStringOriginal;
  end;                                               

function PdParameter.getHint: string;
  begin
  result := self.hint;
  end;

{ ---------------------------------------------------- PdParameterManager }
constructor PdParameterManager.create;
  begin
  inherited create;
  parameters := TListCollection.create;
  end;

destructor PdParameterManager.destroy;
	begin
  parameters.free;
  parameters := nil;
  inherited destroy;
  end;

procedure PdParameterManager.addParameter(newParameter: PdParameter);
	begin
  if newParameter <> nil then
    parameters.add(newParameter);
  end;

procedure PdParameterManager.addParameterForSection(sectionName: string; orthogonalSectionName: string;
    newParameter: PdParameter);
  var
    section: PdSection;
  begin
  section := nil;
  section := domain.sectionManager.sectionForName(sectionName);
  if section = nil then
    begin
    section := domain.sectionManager.addSection(sectionName);
    { the 'no section' section is hidden from the parameters window but we still want the section for writing out }
    section.showInParametersWindow := not (upperCase(sectionName) = upperCase('(no section)'));
    end;
  self.addParameter(newParameter);
  if section <> nil then
    begin
    newParameter.originalSectionName := sectionName;
    section.addSectionItem(newParameter.fieldNumber);
    end;
  // v2.1 orthogonal sections
  // limitation - each param can only have one orthogonal section specified
  // if want to add more later, use delimiter and parse
  orthogonalSectionName := trim(orthogonalSectionName);
  if orthogonalSectionName <> '' then
    begin
    section := nil;
    section := domain.sectionManager.sectionForName(orthogonalSectionName);
    if section = nil then
      begin
      section := domain.sectionManager.addOrthogonalSection(orthogonalSectionName);
      section.showInParametersWindow := true;
      section.isOrthogonal := true;
      end;
    if section <> nil then
      section.addSectionItem(newParameter.fieldNumber);
    end;
  end;

class function PdParameterManager.fieldTypeName(fieldType: integer): string;
  begin
  case fieldType of
    kFieldUndefined: result := 'Undefined';
    kFieldFloat: result := 'Single';
    kFieldSmallint: result := 'Smallint';
    kFieldColor: result := 'Color';
    kFieldBoolean: result := 'Boolean';
    kFieldThreeDObject: result := '3D object';
    kFieldEnumeratedList: result := 'List of choices';
    kFieldLongint: result := 'Longint';
    else result := 'not in list';
    end;
  end;

class function PdParameterManager.indexTypeName(indexType: integer): string;
  begin
  case indexType of
    kIndexTypeUndefined: result := 'Undefined';
    kIndexTypeNone: result := 'None';
    kIndexTypeSCurve: result := 'S curve';
    else result := 'not in list';
    end;
  end;

function PdParameterManager.parameterForIndex(index: longint): PdParameter;
	begin
  result := parameters.items[index];
  end;

procedure PdParameterManager.clearParameters;
  begin
  parameters.clear;
  end;

procedure PdParameterManager.setAllReadFlagsToFalse;
  var
    i: smallint;
  begin
  if parameters.count > 0 then
    for i := 0 to parameters.count - 1 do
      PdParameter(parameters.items[i]).valueHasBeenReadForCurrentPlant := false;
  end;

function PdParameterManager.parameterForFieldNumber(fieldNumber: longint): PdParameter;
  var
    i: smallint;
    parameter: PdParameter;
  begin
  result := nil;
  if parameters.count > 0 then
    for i := 0 to parameters.count - 1 do
      begin
      parameter := PdParameter(parameters.items[i]);
      if fieldNumber = parameter.fieldNumber then
        begin
        result := parameter;
        exit;
        end;
      end;
  raise Exception.create('Problem: Parameter not found for field number ' + intToStr(fieldNumber)
      + ' in method PdParameterManager.parameterForFieldNumber.');
  end;

function PdParameterManager.parameterForFieldID(fieldID: string): PdParameter;
  var
    i: smallint;
    parameter: PdParameter;
  begin
  result := nil;
  if parameters.count > 0 then
    for i := 0 to parameters.count - 1 do
      begin
      parameter := PdParameter(parameters.items[i]);
      if trim(upperCase(fieldID)) = trim(upperCase(parameter.fieldID)) then
        begin
        result := parameter;
        exit;
        end;
      end;
  raise Exception.create('Problem: Parameter not found for field ID ' + fieldID
      + ' in method PdParameterManager.parameterForFieldID.');
  end;

function PdParameterManager.parameterForName(name: string): PdParameter;
  var
    i: smallint;
    parameter: PdParameter;
  begin
  result := nil;
  if parameters.count > 0 then
    for i := 0 to parameters.count - 1 do
      begin
      parameter := PdParameter(parameters.items[i]);
      if trim(upperCase(name)) = trim(upperCase(parameter.name)) then
        begin
        result := parameter;
        exit;
        end;
      end;
  raise Exception.create('Problem: Parameter not found for name ' + name
      + ' in method PdParameterManager.parameterForName.');
  end;

function PdParameterManager.parameterIndexForFieldNumber(fieldNumber: longint): integer;
  var
    i: smallint;
    parameter: PdParameter;
  begin
  result := 0;
  if parameters.count > 0 then
    for i := 0 to parameters.count - 1 do
      begin
      parameter := PdParameter(parameters.items[i]);
      if fieldNumber = parameter.fieldNumber then
        begin
        result := i;
        exit;
        end;
      end;
  raise Exception.create('Problem: Parameter index not found for field number ' + intToStr(fieldNumber)
      + ' in method PdParameterManager.parameterIndexForFieldNumber.');
  end;

function PdParameterManager.parameterIndexForFieldID(fieldID: string): integer;
  var
  i: smallint;
  parameter: PdParameter;
  begin
  result := 0;
  if parameters.count > 0 then
    for i := 0 to parameters.count - 1 do
      begin
      parameter := PdParameter(parameters.items[i]);
      if trim(upperCase(fieldID)) = trim(upperCase(parameter.fieldID)) then 
        begin
        result := i;
        exit;
        end;
      end;
  raise Exception.create('Problem: Parameter index not found for field ID ' + fieldID
      + ' in method PdParameterManager.parameterIndexForFieldID.');
  end;

procedure PdParameterManager.makeParameters;
  begin
  CreateParameters(self);
  CreateParameters_SecondHalf(self);
  end;

end.

File: for-olpc-python/uparams.py
# unit Uparams

import ucollect
import umakepm

"""
from conversion_common import *
import uunits
import usection
import usupport
import uexcept
import delphi_compatability
"""

# const
kFieldUndefined = 0
kFieldFloat = 1
kFieldSmallint = 2
kFieldColor = 3
kFieldBoolean = 4
kFieldThreeDObject = 5
kFieldEnumeratedList = 6
kFieldHeader = 7
kFieldLongint = 8

kIndexTypeUndefined = 0
kIndexTypeNone = 1
kIndexTypeSCurve = 3

kTransferTypeUndefined = 0
kTransferTypeMFD = 1
kTransferTypeGetSetSCurve = 2
kTransferTypeObject3D = 3
kNotArray = -1

# FieldType 
#note: the longint type is NOT for use for parameters (in param panels or breeding).
#     if you want to use a longint for a parameter, you have to
#     update the breeding method and the method that creates param panels. the smallint panel will not
#     work correctly with a longint as it is now (but you could subclass it if you need one).
# IndexType - for arrays 
# transferType - how data is transferred to and from model object 
#mfd is short for MoveFieldData
#must be whole string to allow for tdos
# v2.1
# ---------------------------------------------------------- PdParameter 

class PdParameter:
    def __init__(self):
        self.fieldNumber = 0
        self.fieldID = ""
        self.name = ""
        self.fieldType = 0
        self.indexType = 0
        self.unitSet = 0
        self.unitModel = 0
        self.unitMetric = 0
        self.unitEnglish = 0
        self.lowerBoundOriginal = 0.0
        self.upperBoundOriginal = 0.0
        self.defaultValueStringOriginal = ""
        self.cannotOverride = False
        self.isOverridden = False
        self.lowerBoundOverride = 0.0
        self.upperBoundOverride = 0.0
        self.defaultValueStringOverride = ""
        self.regrow = False
        self.readOnly = False
        self.accessString = ""
        self.transferType = 0
        self.hint = ""
        self.valueHasBeenReadForCurrentPlant = False
        self.collapsed = True
        self.originalSectionName = ""
    
    def make(self, aFieldNumber, aFieldID, aName, aFieldType, anIndexType, aUnitSet, aUnitModel, aUnitMetric, aUnitEnglish, aLowerBound, anUpperBound, aDefaultValueString, aRegrow, aReadOnly, anAccessString, aTransferType, aHint):
        self.fieldNumber = aFieldNumber
        self.fieldID = aFieldID[:80]
        self.name = aName[:80]
        self.fieldType = aFieldType
        # v2.0 headers can have hints
        self.hint = aHint
        if self.fieldType == kFieldHeader:
            return self
        self.indexType = anIndexType
        self.unitSet = aUnitSet
        self.unitModel = aUnitModel
        self.unitMetric = aUnitMetric
        self.unitEnglish = aUnitEnglish
        self.lowerBoundOriginal = aLowerBound
        self.upperBoundOriginal = anUpperBound
        self.defaultValueStringOriginal = aDefaultValueString
        self.regrow = aRegrow
        self.readOnly = aReadOnly
        self.accessString = anAccessString[:80]
        self.transferType = aTransferType
        #hint := aHint;
        return self
    
    def indexCount(self):
        result = 1
        if self.indexType == kIndexTypeUndefined:
            result = 1
        elif self.indexType == kIndexTypeNone:
            result = 1
        elif self.indexType == kIndexTypeSCurve:
            result = 4
        else :
            raise GeneralException.create("Problem: Unexpected index type in method PdParameter.indexCount.")
        return result
    
    def getName(self):
        result = self.name
        return result
    
    def setName(self, newName):
        self.name = UNRESOLVED.copy(newName, 1, 80)
    
    def lowerBound(self):
        if (not self.cannotOverride) and (self.isOverridden):
            result = self.lowerBoundOverride
        else:
            result = self.lowerBoundOriginal
        return result
    
    def upperBound(self):
        if (not self.cannotOverride) and (self.isOverridden):
            result = self.upperBoundOverride
        else:
            result = self.upperBoundOriginal
        return result
    
    def defaultValueString(self):
        if (not self.cannotOverride) and (self.isOverridden):
            result = self.defaultValueStringOverride
        else:
            result = self.defaultValueStringOriginal
        return result
    
    def getHint(self):
        result = self.hint
        return result
    
# ---------------------------------------------------- PdParameterManager 
class PdParameterManager:
    def __init__(self):
        self.parameters = ucollect.TListCollection()
    
    def addParameter(self, newParameter):
        if newParameter != None:
            self.parameters.Add(newParameter)

    def addParameterForSection(self, sectionName, orthogonalSectionName, newParameter):
        # import done here due to circularity issues
        import udomain
        section = udomain.domain.sectionManager.sectionForName(sectionName)
        if section == None:
            section = udomain.domain.sectionManager.addSection(sectionName)
            # the 'no section' section is hidden from the parameters window but we still want the section for writing out 
            section.showInParametersWindow = not (sectionName.upper() == "(no section)".upper())
        self.addParameter(newParameter)
        if section != None:
            newParameter.originalSectionName = sectionName
            section.addSectionItem(newParameter.fieldNumber)
        # v2.1 orthogonal sections
        # limitation - each param can only have one orthogonal section specified
        # if want to add more later, use delimiter and parse
        orthogonalSectionName = orthogonalSectionName.strip()
        if orthogonalSectionName != "":
            section = udomain.domain.sectionManager.sectionForName(orthogonalSectionName)
            if section == None:
                section = udomain.domain.sectionManager.addOrthogonalSection(orthogonalSectionName)
                section.showInParametersWindow = True
                section.isOrthogonal = True
            if section != None:
                section.addSectionItem(newParameter.fieldNumber)
    
    def fieldTypeName(self, fieldType):
        if fieldType == kFieldUndefined:
            result = "Undefined"
        elif fieldType == kFieldFloat:
            result = "Single"
        elif fieldType == kFieldSmallint:
            result = "Smallint"
        elif fieldType == kFieldColor:
            result = "Color"
        elif fieldType == kFieldBoolean:
            result = "Boolean"
        elif fieldType == kFieldThreeDObject:
            result = "3D object"
        elif fieldType == kFieldEnumeratedList:
            result = "List of choices"
        elif fieldType == kFieldLongint:
            result = "Longint"
        else:
            result = "not in list"
        return result
    
    def indexTypeName(self, indexType):
        if indexType == kIndexTypeUndefined:
            result = "Undefined"
        elif indexType == kIndexTypeNone:
            result = "None"
        elif indexType == kIndexTypeSCurve:
            result = "S curve"
        else:
            result = "not in list"
        return result
    
    def parameterForIndex(self, index):
        result = self.parameters[index]
        return result
    
    def clearParameters(self):
        self.parameters.clear()
    
    def setAllReadFlagsToFalse(self):
        for parameter in self.parameters:
            parameter.valueHasBeenReadForCurrentPlant = False
    
    def parameterForFieldNumber(self, fieldNumber):
        result = None
        for parameter in self.parameters:
            if fieldNumber == parameter.fieldNumber:
                result = parameter
                return result
        raise GeneralException.create("Problem: Parameter not found for field number " + IntToStr(fieldNumber) + " in method PdParameterManager.parameterForFieldNumber.")
        return result
    
    def parameterForFieldID(self, fieldID):
        result = None
        for parameter in self.parameters:
            if trim(uppercase(fieldID)) == trim(uppercase(parameter.fieldID)):
                result = parameter
                return result
        raise GeneralException.create("Problem: Parameter not found for field ID " + fieldID + " in method PdParameterManager.parameterForFieldID.")
        return result
    
    def parameterForName(self, name):
        result = None
        for parameter in self.parameters:
            if trim(uppercase(name)) == trim(uppercase(parameter.name)):
                result = parameter
                return result
        raise GeneralException.create("Problem: Parameter not found for name " + name + " in method PdParameterManager.parameterForName.")
        return result
    
    def parameterIndexForFieldNumber(self, fieldNumber):
        result = 0
        for i in range(0, len(self.parameters)):
            parameter = self.parameters[i]
            if fieldNumber == parameter.fieldNumber:
                result = i
                return result
        raise GeneralException.create("Problem: Parameter index not found for field number " + IntToStr(fieldNumber) + " in method PdParameterManager.parameterIndexForFieldNumber.")
        return result
    
    def parameterIndexForFieldID(self, fieldID):
        result = 0
        for i in range(0, len(self.parameters)):
            parameter = self.parameters[i]
            if trim(uppercase(fieldID)) == trim(uppercase(parameter.fieldID)):
                result = i
                return result
        raise GeneralException.create("Problem: Parameter index not found for field ID " + fieldID + " in method PdParameterManager.parameterIndexForFieldID.")
        return result
    
    def makeParameters(self):
        umakepm.CreateParameters(self)
    
